Added cron features to nightly scheduler

vx-i opened the pull request:
I would like to add some cron features to nightly scheduler, like last day of month or 3rd monday of month. Maybe I need help with unit tests.

## Contributor Checklist:

* [ ] I have updated the unit tests
* [x] I have created a file in the `newsfragments` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -67,18 +97,58 @@ def __init__(self, expr_format, start_time=time()):
             while e_list:
                 e = e_list.pop()
+                nth = None
+                special_dow_rem = special_dow_re.match(str(e))
```
Won't `e` be already a string?

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -16,12 +18,29 @@
 only_int_re = re.compile(r'^\d+$')
 any_int_re = re.compile(r'^\d+')
+
+M_ALPHAS = {'jan': 1, 'feb': 2, 'mar': 3, 'apr': 4, 'may': 5, 'jun': 6,
```
MONTH_ALPHAS

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -16,12 +18,29 @@
+
+M_ALPHAS = {'jan': 1, 'feb': 2, 'mar': 3, 'apr': 4, 'may': 5, 'jun': 6,
+            'jul': 7, 'aug': 8, 'sep': 9, 'oct': 10, 'nov': 11, 'dec': 12}
+DOW_ALPHAS = {'sun': 0, 'mon': 1, 'tue': 2, 'wed': 3, 'thu': 4, 'fri': 5, 'sat': 6}
```
DAY_OF_WEEK_ALPHAS

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -16,12 +18,29 @@
+del i
+step_search_re = re.compile(r'^([^-]+)-([^-/]+)(/(\d+))?$')
+
+WEEKDAYS = '|'.join(DOW_ALPHAS.keys())
```
WEEKDAYS_RE_PART

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -16,12 +18,29 @@
+step_search_re = re.compile(r'^([^-]+)-([^-/]+)(/(\d+))?$')
+
+WEEKDAYS = '|'.join(DOW_ALPHAS.keys())
+MONTHS = '|'.join(M_ALPHAS.keys())
```
MONTHS_RE_PART

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -67,18 +97,58 @@ def __init__(self, expr_format, start_time=time()):
+                special_dow_rem = special_dow_re.match(str(e))
+
+                # Handle special case in the day of week expression: 2#3, l3
+                if i == 4 and special_dow_rem:
```
Let's not use `4` as a special constant, but introduce names for them, e.g. `cron_min_index`, `cron_day_of_month_index`, etc.

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -67,18 +97,58 @@ def __init__(self, expr_format, start_time=time()):
+                # Handle special case in the day of week expression: 2#3, l3
+                if i == 4 and special_dow_rem:
+                    g = special_dow_rem.groupdict()
+                    he, last = g.get('he', ''), g.get('last', '')
```
Splitting this into two lines would be easier to understand.

p12tic replied:
Also, `he` and `last` need a more complete names to describe what they mean.

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -35,12 +54,14 @@ class croniter:
-        {'sun': 0, 'mon': 1, 'tue': 2, 'wed': 3, 'thu': 4, 'fri': 5, 'sat': 0},
+        {},  # 0: min
+        {},  # 1: hour
+        {"l": "l"},  # 2: day of month
```
Day of week?

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -16,12 +18,29 @@
 star_or_int_re = re.compile(r'^(\d+|\*)$')
+special_dow_re = re.compile(
+    (r'^(?P<pre>((?P<he>(({WEEKDAYS})(-({WEEKDAYS}))?)').format(WEEKDAYS=WEEKDAYS) +
+    (r'|(({MONTHS})(-({MONTHS}))?)|\w+)#)|l)(?P<last>\d+)$').format(MONTHS=MONTHS)
```
We need a comment of what this regex is trying to accomplish (e.g. what would be the accepted inputs, etc.). Otherwise even if I approve and merge the PR, down the line someone else or even me myself will think that the regex is unnecessarily too complex and simplify it breaking some use cases.

p12tic commented on master/docs/manual/configuration/schedulers.rst:
```diff
@@ -641,7 +641,8 @@ The full list of parameters is:
 
     The day of the month to start a build.
-    This defaults to ``*``, meaning every day.
+    This defaults to ``*``, meaning every day or l for last day.
```
Let's use `L` because the matcher is not case sensitive anyway and `L` is more different from `1`, thus easier to read.

p12tic replied:
Also needs quoting.

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -67,18 +97,58 @@ def __init__(self, expr_format, start_time=time()):
+                            assert 1 <= nth <= 5
+                        except (KeyError, ValueError, AssertionError) as e:
+                            raise ValueError(
+                                "[{expr_format}] is not acceptable.  Invalid day_of_week "
```
I think we miss `f` here to make it `f-string`.

p12tic commented on master/buildbot/util/croniter.py:
```diff
@@ -35,12 +54,14 @@ class croniter:
+        {"l": "l"},  # 2: day of month
+        # 3: month
+        copy.deepcopy(M_ALPHAS),
+        # 4: day of month
```
Day of week?

p12tic commented:
Thanks for the contribution. I'm now wondering - maybe just replacing the current code with a dependency on the `croniter` python package would get all the features that we want without maintaining the same code base twice?

vx-i commented:
> Thanks for the contribution. I'm now wondering - maybe just replacing the current code with a dependency on the `croniter` python package would get all the features that we want without maintaining the same code base twice?

I don't know why the decision was made at the time, but I'm sure it was for a reason. Fewer dependencies means more stable code.

p12tic commented:
> I don't know why the decision was made at the time, but I'm sure it was for a reason. Fewer dependencies means more stable code.

Yes, but in this case we're duplicating the effort by copy-pasting the code. I think a better investment would be to use croniter as a dependency.

vx-i commented:
> Yes, but in this case we're duplicating the effort by copy-pasting the code. I think a better investment would be to use croniter as a dependency.

I would like to disagree on this topic. I just used the croniter code as a starting point. Extracted only relevant parts, refactored code and made it more readable, tested inside Buildbot, added documentation. If you decide to use croniter as an external dependency you skip all my work, delete old internal croniter code, etc., etc. Additionally you add the risk to break things as external API's may change anytime.

vx-i commented:
@p12tic What should we do? Any plans on this topic? To accept PR in the short term and switch to croniter in the medium term?

p12tic commented:
@vx-i Thanks for persisting. Your proposal makes sense as a compromise.
